### PR TITLE
[T3A4][F14-A2]Tan Li Hao

### PR DESCRIPTION
**Pull request by LiHaoTan**
I believe this exercise does not make sense at all because there is no reason to
handle file deleted situation manually when there are better abstractions that does that or in a related way.

In fact, the code already handle it in the way that make sense for its context, i.e. files are always rewritten instead of appending so it does not really matter if files get deleted. FileWriter will throw IOException if if the file exists but is a directory rather than a regular file, does not exist but
cannot be created, or cannot be opened for any other reason.

However, there are situations where files go missing and that can cause issues and it is going to be a headache to handle every single scenario. While I understand that we are focusing on learning OOP in this module so we should not be using databases but I believe the tradeoff is very huge. Using a database ensures ACID and we do not even have to worry about any such issues (though I think some Java APIs offer atomicity).

In the code I have chose to just check if the file exists before saving instead of checking if file is deleted. This is because to truly check if a file is deleted I need to use java.nio or non built-in libraries, i.e., http://docs.oracle.com/javase/7/docs/api/java/nio/file/WatchService.html. However, the code uses JAXB which also requires the use of java.io, which I feel has not been abstracted well and looks very messy (though maybe there is no better way, I don’t know enough to talk about this). Since I do not understand how the IO is handled, do not want to clutter the code further and since the problem is ill defined in the first place because of the relation with data storage, **I will try to just focus on the point of the exercise which is exceptions.**

### CODE REVIEW

**Comment by LiHaoTan** [comment]
Ready for review

**Comment by LiHaoTan** [comment]
**TLDR: I will try to just focus on the point of the exercise which is exceptions.**

File: src/seedu/addressbook/storage/StorageFile.java [hunk 1]
```diff
@@ -136,6 +144,12 @@ public AddressBook load() throws StorageOperationException {
         // create empty file if not found
         } catch (FileNotFoundException fnfe) {
             final AddressBook empty = new AddressBook();
+            // XXX hack to create an empty file since we now do not allow non existent files
+            try {
+                Files.createFile(path);
+            } catch (IOException e) {
+                throw new StorageOperationException("Cannot create file!");
+            }
```
Comment (chanjunweimy) [inline src/seedu/addressbook/storage/StorageFile.java]: while I like your idea of creating a file when the file is not found, it still makes sense for the program to throw an exception. For example, the program might want to tell the user that the previous file is not found, thus need to create a new file. Of course we can do it by normal return boolean or string, a more neat way in OOP is to throw an exception to tell the one who is calling the function: hey, the file is missing but I recover  it, can you do something about this? Then the caller would say ok, I will ask the UI to display the message. ^^

Comment (chanjunweimy) [inline src/seedu/addressbook/storage/StorageFile.java]: But this method does not solve the missing file problem while saving (what if the user deletes the file while using the program)

File: src/seedu/addressbook/storage/StorageFile.java [hunk 2]
```diff
@@ -92,6 +93,13 @@ private static boolean isValidPath(Path filePath) {
      * @throws StorageOperationException if there were errors converting and/or storing data to file.
      */
     public void save(AddressBook addressBook) throws StorageOperationException {
+        
+        // XXX handling of file deleted situation using exceptions.
+        // I think this is a very bad idea. Instead of handling file deleted situation, the
+        // following code will just check if the file exists before saving
+        if (!Files.exists(path)) {
+            throw new StorageOperationException("File does not exist!");
+        }
```
Comment (chanjunweimy) [inline src/seedu/addressbook/storage/StorageFile.java]: checking if the file exists before saving is still make sense because the file might corrupt while the user is editing the program.  Like what you have pointed out, this would degrade the user experience, could you think of other better way in solving it?

Note that, we can also use exception as a warning, instead of stopping the operation ^^

**Comment by chanjunweimy** [comment]
Oh ya, checking if a file exists is a right approach. Good job ^^

File: src/seedu/addressbook/storage/StorageFile.java (hunk 2 above)
Comment (LiHaoTan) [inline src/seedu/addressbook/storage/StorageFile.java]: Oh I did not realise two things based on what you said.
1. Although the program already handles the situation when file does not exist, it makes sense to inform the user that the file might have gone missing and so they can perhaps investigate if it is due to something like corruption because of hard disk failure.
2. The program handle the situation in  a way that is not clear in the code, perhaps it will also be better to spell out that the program handles a missing file situation.
3. I should have handled the exception so the program does not stop.

File: src/seedu/addressbook/storage/StorageFile.java (hunk 1 above)
Comment (LiHaoTan) [inline src/seedu/addressbook/storage/StorageFile.java]: Based on what I have learned from
https://github.com/nus-cs2103-AY1617S2/addressbook-level2/pull/431#discussion_r99479917

This means that the program no longer stops if the program starts with no file at the start, but perhaps I might still want to create an empty file before attempting to save so I can handle the program **start** file missing situation vs program **executing** missing situation separately

Comment (chanjunweimy) [inline src/seedu/addressbook/storage/StorageFile.java]: @LiHaoTan yup, you got it right ^^ you can still create file when you detect there is no file at the beginning though ^^

File: src/seedu/addressbook/storage/StorageFile.java (hunk 2 above)
Comment (chanjunweimy) [inline src/seedu/addressbook/storage/StorageFile.java]: @LiHaoTan yup ^^
